Rearrange the way PDO's are mapped and added load OD from file to node

af-silva opened the pull request:
Rearrange the way PDO's are mapped and also put them in the same package "pdo". I follow the same approach used with the SDO's and created the RPDO and TPDO objects (specializations of the base class PDO), this way I think it is more clear the access of the objects instead of the PDO.rx[1] we use the TPDO[1] to access the Transmit PDO 1. To achieve this I have to refactor some code.

In the class remote I added the functionality to load the values stored in the EDS file. I don't load the default values, only the parameter values defined by the user, I think is safer this way since the default values should not be touch by the user in the file since they are the values loaded by the manufacture.

I'm open for suggestions, this is still work in progress but I would appreciate some input from you guys @christiansandberg @TobiasPleyer.

@TobiasPleyer I'm looking at some code you have done and I will try to used it in this implementation.

This is being tested with real hardware and so far so good.

Thanks.

christiansandberg commented on canopen/node/remote.py:
```diff
@@ -16,22 +24,24 @@ class RemoteNode(BaseNode):
     """
 
-    def __init__(self, node_id, object_dictionary):
+    def __init__(self, node_id, object_dictionary, load_od=False):
```
Add this param to docs.

af-silva replied:
Done

christiansandberg commented on canopen/node/remote.py:
```diff
@@ -16,22 +24,24 @@ class RemoteNode(BaseNode):
         super(RemoteNode, self).__init__(node_id, object_dictionary)
 
-        #: Enable WORKAROUND for reversed PDO mapping entries
-        self.curtis_hack = False
```
Where did this go?

af-silva replied:
Reverted.

christiansandberg commented on canopen/objectdictionary/eds.py:
```diff
@@ -98,6 +98,10 @@ def import_eds(source, node_id):
 
 def import_from_node(node_id, network):
+    """ Download the configuration from the remote node
+    :param hex node_id: Identifier of the node
```
What is hex? Should be int.

af-silva replied:
Done.

af-silva commented on canopen/objectdictionary/eds.py:
```diff
@@ -154,6 +165,22 @@ def build_variable(eds, section, node_id, index, subindex=0):
+            
+            if var.data_type in objectdictionary.DATA_TYPES:
+                var.value = parameter_value
+            elif var.data_type in objectdictionary.DATA_TYPES:
```
Yes, you are right! My mistake.

af-silva commented on canopen/pdo/base.py:
```diff
@@ -146,30 +138,30 @@ def __init__(self, pdo_node, com_record, map_array):
+        # : If this map is valid
         self.enabled = False
-        #: COB-ID for this PDO
+        #: COB-ID for this PdoBase
```
It was, and yes, but you are right there is no need of renaming in the comments PDO do the class name PdoBase, so I have reverted this.

af-silva commented on canopen/example/simple_ds402_node.py:
```diff
@@ -0,0 +1,185 @@
+import canopen
```
Ok, I have moved this to the root of the project

af-silva commented on canopen/nmt.py:
```diff
@@ -51,6 +51,7 @@ def __init__(self, node_id):
         self.id = node_id
         self.network = None
         self._state = 0
+        self._node_guarding_producer = None
```
Ok

af-silva commented:
@christiansandberg Sorry to bother but I would appreciate your input with some questions that I have, I will try to explained my best. 

First one: I was trying the follow some of @TobiasPleyer implementation in is fork but I was a bit confused with the way the PDO's were being handle, since it seems he was replicating functionality already implemented in the **Map** class. My current implementation simplify a bit the access (or at least the understanding of what we are accessing), but if I understand correctly, now we have the RPDO and TPDO that are containers to a **Maps** instance that is a collection of **Map** objects that representes each of the communication channels configured (e.i., RPDO[1], RPDO[2], TPDO[1], ...) and there is already a mechanism for listening the arrival of messages in the Map class trough the function **on_message** that process all the configured _callbacks_ added trough the **add_callback** function. So, if this is the way things work I think it would be nice to implement an "automatic" mechanism in the local node to handle the arrival and dispatch of messages instead of instantiate or extend the class and implement custom callbacks to handle the values stored in the OD. I'm doing this now I just want to make sure i'm on the right path.

Second one: In the arrival/dispatch of messages on the **base** class of the PDO it seems the variables used to configure the communication channel **trans_type**, **inhibit_time**, **event_timer**, **sync_start_value** are only being stored and not used. The main purpose of this variables is to control the way communication is done (e.g., at every sync, RTR only, event driven, etc), so my question is if this control should be implemented in the node (i.e., local) or directly in the **on_message** method?
To implement this we need to take in consideration the need to access how many times the sync has occured... 

Any input would be great, 

Thanks in advance.

christiansandberg commented:
Do you mean that we should store the latest received value in the Object Dictionary? That sounds like a good idea. It could also apply to SDOs.

The PDO configuration is not used yet but should definitely be used for the slave node. I think we should add a SyncConsumer that can notify PDOs on SYNC messages.

I've been thinking a bit about the removal of the `node.pdo` member. Now the user will have to `read()`, `save()` and `load_configuration()` twice. Also the `export()` can only save a file for either transmit or receive PDOs but not both. Since it breaks backwards compatibility I was planning to add a wrapper class anyway, so what do you think about it? Would it make more sense to rename the `pdo.tx` to `pdo.tpdo` for example?

af-silva commented:
Ok, lets tackle this. 

> Do you mean that we should store the latest received value in the Object Dictionary?

Yes, but I'm doing this already, the `ParameterValue` in the EDS file that is loaded by the `load_configuration()` set the `value` variable that is stored in the `ObjectDictionary`. So this variable can be used to hold configuration values loaded by the user but also store values in objects like `0x60FF` (Target velocity) or `0x6064` (Position actual value), what we have to handle are the permissions and types of access. For instance, a local node may required the need to report the Position Actual Value, so in this case we need to have a mechanism to enable the application to set the `value` variable that then is access through a PDO or SDO.

> The PDO configuration is not used yet but should definitely be used for the slave node. I think we should add a SyncConsumer that can notify PDOs on SYNC messages.

I agree, this is what I'm trying to figure it out. The SYNC message, depending on the PDO configuration, can trigger the values to take effect but this is only true (in my point of view) for the local node, since the remote node needs only to handle requests, responses and configuration. So I was looking at the `LocalNode` class implementation to add the mechanism that enables the PDOs to be fired depending on the SYNC message configuration of the PDO.

>I've been thinking a bit about the removal of the node.pdo member. Now the user will have to read(), save() and load_configuration() twice. Also the export() can only save a file for either transmit or receive PDOs but not both. Since it breaks backwards compatibility I was planning to add a wrapper class anyway, so what do you think about it? Would it make more sense to rename the pdo.tx to pdo.tpdo for example?

I don't think I have fully understand what you mean. The `load_configuration()` only loads the configuration from the EDS file to the remote node (slave) for all the objects present in the OD that have the __`ParameterValue`__ variable set. What I did was create two wrappers the `TPDO` and `RPDO` that the only thing they do is to instantiate the mapping (`map`) for the transmit and receive PDO. This way the user can call the objects like `node.rpdo` or `node.tpdo` instead of the `node.pdo.tx` or `node.pdo.tpdo`, I think it makes more sense and it is more readable, if the user wants to access the **RPDO 1** then `node.rpdo[1]`. If I understand, we could remove the `read() `and `save()` from the `PdoBase` class since they are only reading/saving all the configured RPDO and TPDO but this can also be access by each member like so `node.rpdo[1].read()`. I never used the export and to be honest I don't understand it's need.

I'm trying to not break backwards compatibility, I just want to help and try to simplified the use of the API since i'm building my solution around it. I will for now focus my attention in the remote node (since it's my primary goal, I'm trying to control nodes in the can network, and not emulating one). That being said, what do you think it's the best path to follow??

I have improve the access to the PDOs (I think xD ) and have implement the load configuration for the remote node (slave) (it's working ok).
Now I will try to improve the ds402 wrapper to control the nodes and create a interface to enable communication with ROS nodes, but this will be outside the scope of this API.

So, what are your suggestions? 

Thanks

christiansandberg commented on canopen/pdo/base.py:
```diff
@@ -4,59 +4,51 @@
-        raise KeyError("%s was not found in any map" % key)
+        if isinstance(key, int):
+            return self.map[key]
+        raise KeyError("PDO: {0} was not found in any map".format(key))
```
The idea of being able to use a variable name as key is to allow the code to not care about in which PDO the variable happens to be mapped to.

af-silva replied:
fix

christiansandberg commented on canopen/node/remote.py:
```diff
@@ -67,3 +83,28 @@ def restore(self, subindex=1):
         """
         self.sdo.download(0x1011, subindex, b"load")
+
+    def load_configuration(self):
```
This method could actually be in the BaseNode class. The only thing is that you can skip the access type check on a local node as writes will always succeed there.

af-silva replied:
I don't know about this, the ``load_configuration()`` is responsible for the upload of the OD (loaded from the .eds file) to the remote node trough SDO's calls, if the node is local the OD has already the values defined since they were stored in the OD when the .eds file was read.

christiansandberg replied:
Yes you're right. I just need to add support for reading ParameterValue before reading DefaultValue in the local node.

af-silva replied:
Yes, if the ``ParamterValue`` is set it is stored in the ``value`` variable, otherwise the ``value`` variable should hold the ``DefaultValue``. I have not implemented this at the OD level, for both **local** and **remote**, because I wanted to load values only defined in the ``ParamterValue`` to the **remote** node and not the default values, because the remote node could already have values defined and I don't want to overwrite this with the default values.

christiansandberg commented:
Sorry, I misunderstood the `load_configuration()` at first. I thought it was for reading PDO configuration from EDS file. But this is also useful, you just have to read the PDO configuration again using SDO.

You can keep the change for accessing PDOs, it just requires a few more lines of code for the user. I will however try to add a backwards compatibility layer for current implementations. I may be overly careful about changing the API, but I think also you will appreciate it once you build something around a library and can trust that it won't break with the next version.

You don't need to solve the SYNC for local nodes if you don't want to. We should probably start with how to distinguish your TPDO and RPDO classes for local and remote nodes. We probably need something like RemoteTPDO, RemoteRPDO, LocalTPDO, LocalRPDO.

af-silva commented on canopen/emcy.py:
```diff
@@ -13,9 +12,9 @@
 
     def __init__(self):
-        #: Log of all received EMCYs for this node
+        # : Log of all received EMCYs for this node
```
Sorry, this was my IDE auto-format

af-silva commented on canopen/lss.py:
```diff
@@ -19,15 +18,15 @@
-CS_IDENTIFY_REMOTE_SLAVE_SERIAL_NUMBER_HIGH = 0x4B      # m -> s
-CS_IDENTIFY_NON_CONFIGURED_REMOTE_SLAVE = 0x4C          # m -> s
-CS_IDENTIFY_SLAVE = 0x4F                                # s -> m
-CS_IDENTIFY_NON_CONFIGURED_SLAVE = 0x50                 # s -> m
```
Again sorry, my IDE autoformat fault xD

af-silva commented on README.rst:
```diff
@@ -96,28 +96,28 @@ Here are some quick examples of what you can do:
-    speed = node.pdo['Velocity actual value'].phys
-    val = node.pdo['Some group.Some subindex'].raw
+    # Read a value from TPDO[1]
+    node.rpdo[1].wait_for_reception()
```
Fix

af-silva commented on doc/pdo.rst:
```diff
@@ -36,35 +36,36 @@ starts at 1, not 0)::
     # Read current PDO configuration
     node.pdo.read()
```
Done

af-silva commented on canopen/pdo/tpdo.py:
```diff
@@ -0,0 +1,21 @@
+        self.subscribers = {}
+        logger.debug('TPDO Map as {0}'.format(len(self.map)))
+
+    def stop(self):
```
Check the new implementation and see if you agree

af-silva commented on canopen/pdo/base.py:
```diff
@@ -4,59 +4,51 @@
-                for var in pdo_map.map:
-                    yield var.name
+        for var in self.map:
+            yield var.name
```
fix

af-silva commented on canopen/nmt.py:
```diff
@@ -165,19 +162,21 @@ def wait_for_bootup(self, timeout=10):
         """Starts the node guarding mechanism.
         :param float period:
-            frequency (in seconds) at which the node guarding should be advertised to the slave node.
+            Frequency (in seconds) at which the node guarding should be advertised to the slave node.
```
done

af-silva commented on doc/pdo.rst:
```diff
@@ -36,35 +36,36 @@ starts at 1, not 0)::
+    node.tpdo.save()
+    node.rpdo.save()
     # Export a database file of PDO configuration
     node.pdo.export('database.dbc')
```
See the new commit, but I don't understand what you mean... sorry

christiansandberg replied:
The database can be used in some CAN analyzer tool to inspect the PDOs. Previously you could export one database per node, but now you can only export one per TPDO or RPDO which is not very practical. What one wants is probably to have one database for the whole CAN bus anyway.

af-silva replied:
Hum ok, in the last commits I think I resolved this issue with the backward compatibility you could call the PDO instance and call the export. I don't know if this is the beast approach but if you have any ideas I can implement. Thanks for the feedback ;)

christiansandberg replied:
Don't worry, I'll take care of it.

af-silva commented on canopen/node/base.py:
```diff
@@ -12,6 +12,22 @@ class BaseNode(object):
 
+    """PDO Transmission types.
+    For more information please refer to http://www.canopensolutions.com/english/about_canopen/pdo.shtml"""
+    PTT = {
```
It was for future use, but I have removed them from now.

af-silva commented on README.rst:
```diff
@@ -96,28 +96,28 @@ Here are some quick examples of what you can do:
-    node.pdo.tx[1].trans_type = 254
-    node.pdo.tx[1].event_timer = 10
-    node.pdo.tx[1].enabled = True
+    node.tpdo.read()
```
done

af-silva commented on canopen/node/remote.py:
```diff
@@ -67,3 +83,28 @@ def restore(self, subindex=1):
+                            logger.info(str(e))
+                        except canopen.SdoAbortedError as e:
+                            # WORKAROUND for broken implementations: the SDO is set but the error
+                            # "Attempt to write a read-only object" is raised any way.
```
This happens with me, I receive the "Attempt to write a read-only object" but the SDO sets the Object with the intended value...

af-silva commented on canopen/node/remote.py:
```diff
@@ -67,3 +83,28 @@ def restore(self, subindex=1):
+                            if e.code != 0x06010002:
+                                # Abort codes other than "Attempt to write a read-only object"
+                                # should still be reported.
+                                print ('[ERROR SETTING object {0}:{1}]  {2}'.format(subobj.index, subobj.subindex, str(e)))
```
done

af-silva commented on canopen/node/remote.py:
```diff
@@ -67,3 +83,28 @@ def restore(self, subindex=1):
+        for obj in self.object_dictionary.values():
+            if isinstance(obj, Record) or isinstance(obj, Array):
+                for subobj in obj.values():
+                    if isinstance(subobj, Variable) and (subobj.access_type == 'rw') and (subobj.value != None) :
```
done

af-silva commented on canopen/node/remote.py:
```diff
@@ -67,3 +83,28 @@ def restore(self, subindex=1):
+    def load_configuration(self):
+        ''' Load the configuration of the node from the object dictionary.'''
+        for obj in self.object_dictionary.values():
+            if isinstance(obj, Record) or isinstance(obj, Array):
```
it's below, I follow your approach ...

af-silva commented on canopen/objectdictionary/eds.py:
```diff
@@ -147,15 +156,32 @@ def build_variable(eds, section, node_id, index, subindex=0):
+                else:
+                    var.value = int(parameter_value, 0)
+        except ValueError:
+            pass
```
see implementation please

af-silva commented:
Added backwards compatibility with the "old" API when accessing the PDOs, see if you agree @christiansandberg. Any suggestion say so. I will go back to the remote handling of the nodes, I'm adding more support for the ds402 interface, for now I can do everything with the remote node (at least until now xD )

christiansandberg commented on canopen/objectdictionary/eds.py:
```diff
@@ -116,7 +118,27 @@ def import_from_node(node_id, network):
+        if '$NODEID+' in value and node_id is not None:
+            var = int(value.replace('$NODEID+', ''), 0) + node_id
+        else:
+            var = int(value, 0)
```
You have to return the value.

af-silva replied:
fixed

christiansandberg commented on canopen/pdo/__init__.py:
```diff
@@ -0,0 +1,62 @@
+        """Stop transmission of all RPDOs.
+        :raise TypeError: Exception is thrown if the node associated with the PDO does not
+        support this function"""
+        if isinstance(canopen.RemoteNode, self.node):
```
The arguments should be switched.

af-silva replied:
fixed

christiansandberg commented on doc/pdo.rst:
```diff
@@ -31,40 +31,43 @@ Examples
 
     # Read current PDO configuration
-    node.pdo.read()
+    node.pdo[4].read()
```
This should be reading both TPDO and RPDO instead.

af-silva replied:
done, calling the ``node.pdo.read()`` again

christiansandberg commented on doc/pdo.rst:
```diff
@@ -73,27 +76,38 @@ starts at 1, not 0)::
+   .. py:attribute:: map
 
-      The :class:`canopen.pdo.Maps` object representing the receive PDO maps.
+      The :class:`canopen.pdo.Mcps` object representing map associated with the instantiated PDO (transmit or receive).
```
Typo Mcps.

af-silva replied:
fixed

christiansandberg commented on doc/pdo.rst:
```diff
@@ -73,27 +76,38 @@ starts at 1, not 0)::
-   .. py:attribute:: tx
 
-      The :class:`canopen.pdo.Maps` object representing the transmit PDO maps.
+.. autoclass:: canopen.pdo.PDO
```
You don't have to add this to the documentation.

af-silva replied:
ok

christiansandberg commented on doc/pdo.rst:
```diff
@@ -73,27 +76,38 @@ starts at 1, not 0)::
+.. autoclass:: canopen.pdo.TPDO
+   :members:
+
+   .. method:: stop()
```
Won't this show up using autodoc like the other methods?

af-silva replied:
I think so. I have it removed from the pdo.rst

christiansandberg commented on examples/simple_ds402_node.py:
```diff
@@ -0,0 +1,160 @@
+        print("Error {0} was found in the log".format(error.raw))
+
+    for node_id in network:
+        print(network[node_id])
```
You're getting this since you are mixing print functions and print statements. You should use print functions only as print statements only work on Python 2.7.

christiansandberg commented on canopen/pdo/__init__.py:
```diff
@@ -0,0 +1,62 @@
+logger = logging.getLogger(__name__)
+
+
+class PDO(PdoBase):
```
Could you also add the `read()` and `save()` methods?

af-silva replied:
The ``read()`` and ``save()`` methods are available from the PdoBase.

af-silva commented:
> There are a few tests that are failing due to the changes.

All the tests are passing! Do you want me to add specific tests to the stuff I have changed??
So now this branch has:

DONE: 
- [x] Full support to the "old" API (backwards compatibility) 
- [x] Added a more simplified way to access the PDOs (e.g., node.TPDO[1]['StatusWord'])
- [x] Added support to upload the OD (loaded from the EDS file)  to the remote CAN node trough SDOs
- [x] Added support for the OD to hold the ``ParameterValue`` that specifies the last value stored in a specific object
- [x] Updated documentation (it needs improvement ??? )

TODO: 
- [ ] Add support to access the configured PDOs in a more agnostic way, using the object ID instead of the NAME (e.g., node.TPDO[1][0x6041])
- [ ] Add more tests

What do you think @christiansandberg? Any feedback??

christiansandberg commented on canopen/pdo/__init__.py:
```diff
@@ -0,0 +1,62 @@
+        super(PDO, self).__init__(node)
+        self.rx = rpdo.map
+        self.tx = tpdo.map
+        self.map = dict(itertools.chain(self.rx.iteritems(), self.tx.iteritems()))
```
`iteritems()` is not available anymore on Python 3. Try with just `items()`.

af-silva replied:
Of course you are right, thanks. I have solved the issue and by doing so added another way to access the PDO! (added some tests to validate this)

Now we can access the PDOs in three forms:

1. Traditional way node.pdo.tx[1]
2. Simplified way node.tpdo[1]
3. New way node.pdo[0x1A00]

I think now everything works as expected, what do you say?

christiansandberg replied:
That should work. Yes I don't think I have anything else to comment on. Not sure why Travis isn't triggering and why there are still conflicts with lss.py. Do you want to merge it now? Would you mind if I squashed your commits?

af-silva replied:
Do what you think is best @christiansandberg, i'm glad I can help.
The conflicts with the file ``canopen/lss.py`` I think is due to a mistake I did when auto formatting the files with my IDE, then the codacy bot was complaining about *trailing white spaces* so I removed them causing the conflict, you can (I think) ignore it. 

I'm finishing a new ds402 profile implementation, based on what was already available but with more functionality.

Then I might release a ROS wrapper around this library our in a separated project with this one as submodule. 

Thanks for all the help @christiansandberg.
